cob_cartesian_controller/src/cartesian_controller_utils.cpp

```diff
@@ -0,0 +1,161 @@
+/*!
+ *****************************************************************
+ * \file
+ *
+ * \note
+ *   Copyright (c) 2015 \n
+ *   Fraunhofer Institute for Manufacturing Engineering
+ *   and Automation (IPA) \n\n
+ *
+ *****************************************************************
+ *
+ * \note
+ *   Project name: care-o-bot
+ * \note
+ *   ROS stack name: cob_control
+ * \note
+ *   ROS package name: cob_cartesian_controller
+ *
+ * \author
+ *   Author: Christoph Mark, email: christoph.mark@ipa.fraunhofer.de / christoph.mark@gmail.com
+ *
+ * \date Date of creation: July, 2015
+ *
+ * \brief
+ *   ...
+ *
+ ****************************************************************/
+
+#include <cob_cartesian_controller/cartesian_controller_utils.h>
+
+
+geometry_msgs::Pose CartesianControllerUtils::getPose(std::string target_frame, std::string source_frame)
+{
+    geometry_msgs::Pose pose;
+    tf::StampedTransform stamped_transform;
+    
+    stamped_transform = getStampedTransform(target_frame, source_frame);
+    
+    tf::pointTFToMsg(stamped_transform.getOrigin(), pose.position);
+    tf::quaternionTFToMsg(stamped_transform.getRotation(), pose.orientation);
+
+    return pose;
+}
+
+tf::StampedTransform CartesianControllerUtils::getStampedTransform(std::string target_frame, std::string source_frame)
+{
+    tf::StampedTransform stamped_transform;
+    bool transform = false;
+    
+    do
+    {
+        try
+        {
+            ros::Time now = ros::Time::now();
+            tf_listener_.waitForTransform(target_frame, source_frame, now, ros::Duration(0.5));
+            tf_listener_.lookupTransform(target_frame, source_frame, now, stamped_transform);
+            transform = true;
+        }
+        catch (tf::TransformException& ex)
+        {
+            ROS_ERROR("%s",ex.what());
+            ros::Duration(0.1).sleep();
+        }
+    }while(!transform);
+    
+    return stamped_transform;
+}
+
+void CartesianControllerUtils::transformPose(const std::string source_frame, const std::string target_frame, const geometry_msgs::Pose pose_in, geometry_msgs::Pose& pose_out)
+{
+    bool transform = false;
+    geometry_msgs::PoseStamped stamped_in, stamped_out;
+    stamped_in.header.frame_id = source_frame;
+    stamped_in.pose = pose_in;
+    
+    do
+    {
+        try
+        {
+            stamped_in.header.stamp = ros::Time::now();
+
+            tf_listener_.waitForTransform(target_frame, source_frame, stamped_in.header.stamp, ros::Duration(0.5));
+            tf_listener_.transformPose(target_frame, stamped_in, stamped_out);
+            pose_out = stamped_out.pose;
+            transform = true;
+        }
+        catch (tf::TransformException& ex)
+        {
+            ROS_ERROR("%s",ex.what());
+            ros::Duration(0.1).sleep();
+        }
+    }while(!transform);
```

Comment:
Worst case: endless loop in case of no transform could be found (as occurred in my case)